scripts/run_mock_server.py

```diff
@@ -0,0 +1,350 @@
+"""Run a mock zwave-js-server instance off of a network state dump."""
+import argparse
+import asyncio
+from collections import defaultdict
+from collections.abc import Hashable
+import json
+import logging
+from typing import DefaultDict, List, Optional, Union
+
+from aiohttp import WSMsgType, web, web_request
+
+from zwave_js_server.client import SIZE_PARSE_JSON_EXECUTOR
+from zwave_js_server.const import MAX_SERVER_SCHEMA_VERSION, MIN_SERVER_SCHEMA_VERSION
+from zwave_js_server.model.version import VersionInfoDataType
+
+DATEFMT = "%Y-%m-%d %H:%M:%S"
+FMT = "%(asctime)s [%(levelname)s] %(message)s"
+
+
+class ExitException(Exception):
+    """Represent an exit error."""
+
+
+# https://stackoverflow.com/a/1151686
+class HashableDict(dict):
+    """Dictionary that can be used as a key in a dictionary."""
+
+    def __key(self) -> tuple:
+        return tuple((k, self[k]) for k in sorted(self))
+
+    def __hash__(self) -> int:
+        return hash(self.__key())
+
+    def __eq__(self, other) -> bool:
+        # pylint: disable=protected-access
+        return self.__key() == other.__key()
+
+
+class MockZwaveJsServer:
+    """
+    Class to represent a mock zwave-js-server instance.
+
+    Only designed to handle one client at a time.
+    """
+
+    def __init__(
+        self,
+        network_state_dump: List[dict],
+        events_to_replay: List[dict],
+        command_responses: DefaultDict[HashableDict, list],
+    ) -> None:
+        """Initialize class."""
+        self.network_state_dump = network_state_dump
+        self.app = web.Application()
+        self.app.add_routes(
+            [
+                web.get("/", self.server_handler),
+                web.post("/replay", self.replay_handler),
+            ]
+        )
+        self.primary_ws_resp: Optional[web.WebSocketResponse] = None
+        self.events_to_replay = events_to_replay
+        self.command_responses = command_responses
+
+    async def send_json(self, data: dict) -> None:
+        """Send JSON."""
+        logging.debug("Sending JSON: %s", data)
+        assert self.primary_ws_resp is not None
+        await self.primary_ws_resp.send_json(data)
+
+    async def send_command_response(
+        self,
+        data: dict,
+        message_id: str,
+    ) -> None:
+        """Send message."""
+        await self.send_json({**data, "messageId": message_id})
+
+    async def send_success_command_response(
+        self,
+        result: Optional[dict],
+        message_id: str,
+    ) -> None:
+        """Send success message."""
+        if result is None:
+            result = {}
+        await self.send_command_response(
+            {"result": result, "type": "result", "success": True}, message_id
+        )
+
+    async def process_record(self, record: dict) -> None:
+        """Process a replay dump record."""
+        if record["record_type"] == "event":
```

Comment:
Should we make a guard clause instead and require the record_type key with value event or command, otherwise raise ExitException? Otherwise there could be a KeyError here perhaps?

Reply:
I improved guards in both locations. With that being said, I think we should ignore any additional lack of guards for now and focus on functionality. We can always add in more guards later and because this is a developer tool, it's on us to make sure that both the method to generate these replay dumps and the server logic work well in tandem.